lib/pid/src/PID.cpp

```diff
@@ -1,492 +1,120 @@
-    tempTauR = 0.0;
-  } else {
-    tempTauR = (1.0 / tauI) * tSample_.count();
-  }
-
-  // For "bumpless transfer" we need to rescale the accumulated error.
-  if (inAuto) {
-    if (tempTauR == 0.0) {
-      accError_ = 0.0;
-    } else {
-      accError_ *= (Kc_ * tauR_) / (Kc * tempTauR);
-    }
-  }
-
-  Kc_   = Kc;
-  tauR_ = tempTauR;
-  tauD_ = tauD / tSample_.count();
+void PID::PID::updateDeadzone(float deadzone) {
+  std::lock_guard<Mutex> lock(m_mutex);
+  m_deadzone = deadzone;
 }
 
-void PID::reset(void) {
-  float scaledBias = 0.0;
-
-  if (usingFeedForward) {
-    scaledBias = (bias_ - outMin_) / outSpan_;
-  } else {
-    scaledBias = (realOutput_ - outMin_) / outSpan_;
-  }
-
-  prevControllerOutput_ = scaledBias;
-  prevProcessVariable_  = (processVariable_ - inMin_) / inSpan_;
-
-  // Clear any error in the integral.
-  accError_ = 0;
-}
-
-void PID::setMode(int mode) {
-  // We were in manual, and we just got set to auto.
-  // Reset the controller internals.
-  if (mode != 0 && !inAuto) {
-    reset();
-  }
-
-  inAuto = (mode != 0);
+float PID::PID::reportProportionalGain() const {
+  std::lock_guard<Mutex> lock(m_mutex);
+  return m_PGain;
 }
 
-void PID::setInterval(std::chrono::duration<float> interval) {
-  if (interval > 0s) {
-    // Convert the time-based tunings to reflect this change.
-    tauR_ *= (interval / tSample_);
-    accError_ *= (tSample_ / interval);
-    tauD_ *= (interval / tSample_);
-    tSample_ = interval;
-  }
-}
-
-void PID::setSetPoint(float sp) {
-  setPoint_ = sp;
+float PID::PID::reportIntegralGain() const {
+  std::lock_guard<Mutex> lock(m_mutex);
+  return m_IGain;
 }
 
-void PID::setProcessValue(float pv) {
-  processVariable_ = pv;
+float PID::PID::reportDerivativeGain() const {
+  std::lock_guard<Mutex> lock(m_mutex);
+  return m_DGain;
 }
 
-void PID::setBias(float bias) {
-  bias_            = bias;
-  usingFeedForward = 1;
+float PID::PID::reportDeadzone() const {
+  std::lock_guard<Mutex> lock(m_mutex);
+  return m_deadzone;
 }
 
-void PID::setDeadZoneError(float error) {
-  deadZoneError_ = error;
+void PID::PID::reset() {
+  std::lock_guard<Mutex> lock(m_mutex);
+  m_IAccumulator = 0;
+  m_pastError    = 0;
+  m_pastPV       = 0;
+  m_timer.stop();
+  m_timer.reset();
 }
 
-void PID::setRealOutput(float realOutput) {
-  realOutput_ = realOutput;
+float PID::PID::computePPath(float error) {
+  // no mutex lock needed since inside compute() only
+  return error * m_PGain;
 }
 
-void PID::setupAutoTune(void *outputPointer, float *inputPointer, int actuatorType) {
-  // set actuator type
-  actuatorType_ = actuatorType;
-
-  // set I/O pointers
-  output_ = outputPointer;
-  input_  = inputPointer;
+float PID::PID::computeIPath(float error, int64_t dt) {
+  // no mutex lock needed since inside compute() only
+  m_IAccumulator += error * dt * m_IGain;
+  m_IAccumulator = std::clamp(m_IAccumulator, static_cast<float>(m_lowerBound), static_cast<float>(m_upperBound));
```

Comment:
1. should provide a way to disable anti-windup.
2. isn't this implemented wrong? should the logic be to not accumulate error when the output is saturated?

Reply:
why would we want to allow wind up?
thats what the std::clamp part is for. from beauregard's blog, the idea is to not allow the pid to ask for a control signal beyond the bounds. so we clamp once here and again once we sum up all the paths

Reply:
1. Because there are many different ways to implement anti windup and they each may have unintended consequences. I think it should default to enabled but there should be a way to disable it just like u have for error on pv

Reply:
re the clamping method: 
Theres 2 things i wanted to bring up.

1. see the comment by Will in the blog post as well as look through this issue on[ github](https://github.com/br3ttb/Arduino-PID-Library/issues/76) about conditional integration. conditional integration is the way I've done it before and i think its more correct. what u have now, the integral term can continue to grow even when output is saturated (due to contributions from i and d term).

2. the clamp in compute() also functions as a saturation block (aside for using it in your "antiwindup") and it should be able to be disabled in config.

Reply:
i think maybe we should chat sometime to discuss this

Reply:
btw avg error is has grown to 8000 after this last commit

Reply:
oof. probs a bug or something?